src/darker/formatters/ruff_formatter.py

```diff
@@ -0,0 +1,198 @@
+        if self.config.get("skip_string_normalization", False):
+            args.append('''--config=format.quote-style="preserve"''')
+        if self.config.get("preview", False):
+            args.append("--preview")
+
+        # The custom handling of empty and all-whitespace files below will be
+        # unnecessary if https://github.com/psf/ruff/pull/2484 lands in Ruff.
+        contents_for_ruff = content.string_with_newline("\n")
+        dst_contents = _ruff_format_stdin(contents_for_ruff, path_from_cwd, args)
+        return TextDocument.from_str(
+            dst_contents,
+            encoding=content.encoding,
+            override_newline=content.newline,
+        )
+
+    def _read_config_file(self, config_path: str) -> None:
+        """Read Ruff configuration from a configuration file."""
+        with Path(config_path).open(mode="rb") as config_file:
+            raw_config = tomllib.load(config_file).get("tool", {}).get("ruff", {})
+        if "line_length" in raw_config:
+            self.config["line_length"] = raw_config["line_length"]
+
+    def _read_cli_args(self, args: Namespace) -> None:
+        return read_black_compatible_cli_args(args, self.config)
+
+    def get_config_path(self) -> str | None:
+        """Get the path of the configuration file."""
+        return self.config.get("config")
+
+    def get_line_length(self) -> int | None:
+        """Get the ``line-length`` configuration option value."""
+        return self.config.get("line_length")
+
+    def get_exclude(self, default: Pattern[str]) -> Pattern[str]:
+        """Get the ``exclude`` configuration option value."""
+        return self.config.get("exclude", default)
+
+    def get_extend_exclude(self) -> Pattern[str] | None:
+        """Get the ``extend_exclude`` configuration option value."""
+        return self.config.get("extend_exclude")
+
+    def get_force_exclude(self) -> Pattern[str] | None:
+        """Get the ``force_exclude`` configuration option value."""
+        return self.config.get("force_exclude")
+
+
+def _get_supported_target_versions() -> set[str]:
+    """Get the supported target versions for Ruff.
+
+    Calls ``ruff config target-version`` as a subprocess, looks for the line looking
+    like ``Type: "py38" | "py39" | "py310"``, and returns the target versions as a set
+    of strings.
+
+    """
+    cmdline = "ruff config target-version"
+    output = run(  # noqa: S603  # nosec
+        cmdline.split(), stdout=PIPE, check=True, text=True
+    ).stdout
+    type_lines = [line for line in output.splitlines() if line.startswith('Type: "py')]
+    if not type_lines:
+        message = f"`{cmdline}` returned no target versions on a 'Type: \"py...' line"
+        raise ConfigurationError(message)
+    quoted_targets = type_lines[0][6:].split(" | ")
+    if any(tgt_ver[0] != '"' or tgt_ver[-1] != '"' for tgt_ver in quoted_targets):
+        message = f"`{cmdline}` returned invalid target versions {type_lines[0]!r}"
```

Comment:
!r at the end I don't understand what it does

Reply:
```python
>>> your_guy = "Arthur Dent"
>>> print(f"Your guy is {your_guy}")
Your guy is Arthur Dent
>>> print(f"Your guy is {your_guy!r}")
Your guy is 'Arthur Dent'
```

From Python's [f-strings] documentation:
>If a conversion is specified, the result of evaluating the expression is converted before formatting. Conversion `'!s'` calls [`str()`](https://docs.python.org/3/library/stdtypes.html#str "str") on the result, `'!r'` calls [`repr()`](https://docs.python.org/3/library/functions.html#repr "repr"), and `'!a'` calls [`ascii()`](https://docs.python.org/3/library/functions.html#ascii "ascii").

[f-strings]: https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals

Reply:
Ok that is pretty handy, thank you for the explanation on this one